Fix in game buttons refresh

thiagorb opened the pull request:
Please help me test this! :smile:

Fixes #1427.

This PR is almost a complete rewrite of the build, manufacture, research and commander interfaces. It was an idea suggested by @past-due to eliminate the issues caused by the "screen refresh".

The previous implementation used the screen refresh to recreate the interface widgets from scratch, every time an event in the game happened that affected the interface, for example: a droid being destroyed, a droid finishing a construction, a research being completed, etc. This was necessary because the previous implementation was binding a widget representing a game object with the object itself during the widget creation, so if the game object was changed, the widget related to that object had to be recreated. Or if a new object was created (like a new builder, or a new factory, or a new research facility), a new button referring to this new object had to be added to the interface.

With the new implementation, the widgets only hold a reference to what position they represent. So, if the player has 3 builders, 3 "builder buttons" will be created, and each of them will remember they are the _i_-th, builder button. In every frame, each button figures out which builder they refer to, check what the builder is doing and update its sub-widgets (like construction progress bar). Whenever a "screen refresh" is triggered by events in the game, the list of builders is recalculated, but the widgets   don't need to be recreated.

It might be possible to improve this implementation removing completely the need to "refresh the screen" relying on game events, if we just recalculate the list of objects on every frame, but I believe we would need to improve the game data structures to be able to query the data more efficiently. A good start would be to store droids and structures mapped by type, so that we can avoid iterating through the entire list of objects.

KJeff01 commented:
Some observations so far:

1. Hovering over a factory widget and then issuing production orders, will start the droid with the same rotation as the factory in the preview.

2. If a menu automatically closes from losing all research labs, for example, this gets output by my terminal:
```
info    |20:25:54: [setHighlightedObject:54] facility should not be nullptr
info    |20:25:54: [setHighlightedObject:54] Assert in Warzone: ../src/hci/research.h:54 ((facility) != nullptr), last script event: ''
```

3. Seems production menu breaks if producing units from offworld. This can be tested by entering cheat mode and then using the "let me win" cheat to go into Beta 2.

thiagorb commented:
I just pushed a fix for the number 3.

AlexTheDacian commented:
Let's start this with: Amazeballs! :partying_face: 

From what I've tested so far, it works as you noted, no more sudden screen refreshing after:
- a unit was built while the player is looking at the **Manufacture tab**
- multiple units have been destroyed
- a research project has been finished while the player is looking at the **Research tab**

Also, it seems that @KJeff01 's **No.3 observation** [_production menu breaking_] was fixed as well, from what I've seen in-game. But so far, I haven't encountered the **No.1 observation** [_rotated unit due to hovering over the factory widget_].
Will test some more for it. But if I don't encounter it, then it's just a glitch somewhere.

Instead I found something else that may affect the folks playing Multiplayer. The ability to **cancel a unit production queue or a research project via right-clicking on the unit/research above the respective Factory/Research Facility** is now missing.
For me, this isn't really a deal-breaker, but as I said before, it may affect the Online players.

Here's an example of the feature both in 3.4.1 and the lack of it in the PR.

**3.4.1 Release**
![](https://cdn.discordapp.com/attachments/685112194614296587/803901782723657768/ezgif.com-gif-maker_6.gif)

**Thiago's PR**
![](https://cdn.discordapp.com/attachments/685112194614296587/803902325382709248/ezgif.com-gif-maker_7.gif)

Looking forward to more stuff from you @thiagorb ! :+1:

thiagorb commented:
@AlexTheDacian, I implemented the function to hold/cancel the production in the manufacture interface. The research cancellation was working fine for me, let me know if you still find any problem.

@KJeff01 I fixed the issue with the last research facility being destroyed. I didn't get the point number 1 though. Is it possible to make a video to show what you mean?

Thanks a lot for the help guys!

KJeff01 commented:
Appears the factory widget is not a factor. It's easy enough to see by right-click canceling production by hovering over the bottom widget and trying again repeatedly with a new template. It looks like the rotation from the right-side menu widget influences the bottom widget initial rotation.

past-due commented:
@KJeff01 Please let me know if the commit I just pushed eliminates your concern.

KJeff01 commented:
@past-due Yep. I can't reproduce it anymore. Real nice touch rotating things back to the default.

past-due reviewed (approved):
🥳
